src/neuralnet/project.cpp

```diff
@@ -0,0 +1,210 @@
+#include "project.h"
+
+#include <algorithm>
+#include <atomic>
+#include <boost/algorithm/string/case_conv.hpp>
+
+using namespace NN;
+
+//!
+//! \brief A mutable representation of a \c Project.
+//!
+//! The internal implementation uses this class to sort containers of \c Project
+//! instances. \c std::sort() cannot sort a collection of \c const objects.
+//!
+struct MutableProject
```

Comment:
I'm working on a way to get rid of `MutableProject`. It mainly exists because I declared the member fields of `Project` as `const`, so `std::sort()` can't operate on the vector because it wants to move the values into memory marked immutable as it rearranges the items. `LowercaseName()` might speed up the sort comparison because each project name is compared multiple times, but we sort the set so infrequently that it doesn't really matter...would like to get rid of the extra struct.

Reply:
This may be an idiotic question but why do the member fields need to be const?

Reply:
They...probably don't. I wanted to guarantee that calling code can't modify the contract data in a `Project`. The interface of `WhitelistSnapshot` seems to prevent this from happening. Need to find some time to verify that. Worst case, we can put the `Project` member fields behind accessor methods like `Name()`, `Url()`, and `Timestamp()`.

Reply:
I would rather do that, and then you can do a sort on the vector.